mlir/lib/IR/AsmPrinter.cpp

```diff
@@ -1553,7 +1570,12 @@ void SSANameState::numberValuesInBlock(Block &block) {
       specialNameBuffer.resize(strlen("arg"));
       specialName << nextArgumentID++;
     }
-    setValueName(arg, specialName.str());
+    if (printerFlags.shouldUseNameLocAsPrefix() && isa<NameLoc>(arg.getLoc())) {
+      auto nameLoc = cast<NameLoc>(arg.getLoc());
+      setValueName(arg, nameLoc.getName());
```

Comment:
Can you share the implementation of these three usages? Seems like they all pretty much have the same pattern.

Also for the NameLoc itself, you likely want to use `arg.getLoc().findInstanceOf<NameLoc>()`, given that in some cases (e.g. DebugInfo) you may not have a top level NameLoc.

Reply:
I tried to factor out the whole conditional but then you end up hiding the `printerFlags.shouldUseNameLocAsPrefix()` in the helper itself. I didn't really like that (would be surprising I think) so I factored out the "get the name from a (possibly) nested `NameLoc`".